modules/context/repo.go

```diff
@@ -618,10 +619,11 @@ func RepoRefByType(refType RepoRefType) macaron.Handler {
 
 			if refType == RepoRefLegacy {
 				// redirect from old URL scheme to new URL scheme
+				unescaped, _ := url.PathUnescape(ctx.Req.URL.String())
 				ctx.Redirect(path.Join(
 					setting.AppSubURL,
-					strings.TrimSuffix(ctx.Req.URL.String(), ctx.Params("*")),
-					ctx.Repo.BranchNameSubURL(),
+					strings.TrimSuffix(unescaped, ctx.Params("*")),
+					(&url.URL{Path: ctx.Repo.BranchNameSubURL()}).String(),
```

Comment:
Can't we just use `url.PathEscape` here instead? Just a thought. Didn't try it myself.

Reply:
When I tried that the last time, I got back into the endless loop that I was trying to fix in the first place. I have no idea what exactly the difference is.

Reply:
When I tried that the last time, I got back into the endless loop that I was trying to fix in the first place. I have no idea what exactly the difference is.

Reply:
I looked into that again:
String(): `branch/s%3Cscript%3Ealert%28%27XSS%27%29;%3C/script%3Es`
PathEscape(): `branch%2Fs%3Cscript%3Ealert%28%27XSS%27%29%3B%3C%2Fscript%3Es`

PathEscape also escapes slashes, which makes sense but is not wanted here because BranchNameSubURL already contains the branch/ prefix.